various-receipes/typescript/test/typescript-example.ts

```diff
@@ -0,0 +1,257 @@
+import * as express from "express";
+import * as nock from "nock";
+import * as request from "supertest";
+import * as sinon from "sinon";
+
+import { initializeWebServer, stopWebServer } from "../../../example-application/api";
+import OrderRepository = require("../../../example-application/data-access/order-repository.js");
+
+let expressApp: express.Application;
+
+beforeAll(async (done) => {
+  // ️️️✅ Best Practice: Place the backend under test within the same process
+  expressApp = await initializeWebServer();
+
+  // ️️️✅ Best Practice: Ensure that this component is isolated by preventing unknown calls
+  nock.disableNetConnect();
+  nock.enableNetConnect('127.0.0.1');
+
+  done();
+});
+
+beforeEach(() => {
+  nock('http://localhost/user/').get(`/1`).reply(200, {
+    id: 1,
+    name: 'John',
+  });
+});
+
+afterEach(() => {
+  nock.cleanAll();
+  sinon.restore();
+});
+
+afterAll(async (done) => {
+  // ️️️✅ Best Practice: Clean-up resources after each run
+  await stopWebServer();
+  nock.enableNetConnect();
+  done();
+});
+
+// ️️️✅ Best Practice: Structure tests
+describe('/api', () => {
+  describe('GET /order', () => {
+    test('When asked for an existing order, Then should retrieve it and receive 200 response', async () => {
+      //Arrange
+      const orderToAdd = {
+        userId: 1,
+        productId: 2,
+        mode: 'approved',
+      };
+      const {
```

Comment:
Challenge - If a person is already using TS, she probably wants to leverage this and get type safety in testing. This variable (orderToAdd) has no type. Can we lightly declare its type without defining a class? Maybe TS can infer from OpenAPI? Maybe the server should also export a client or at least d.ts?

Note that this deals with the core of our mission - It's both about convenience, but also will prevent bugs if the API is typesafe

 Let's leverage the TS advantages without increasing the writing burden. This research will surely boost your TS skills and will be interesting